### PR TITLE
[WIP] Create application controller fixture, offer category model and manager

### PR DESCRIPTION
**Pull request by iamanikeev**

### CODE REVIEW

**Comment by iamanikeev** [comment]
Will add test cases tomorrow

File: pyoffers/models/application.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        return self._call(
+            'findAllOfferCategories',
+            filters=Filter(**kwargs),
+            sort=sort,
+            fields=fields, single_result=False
+        )
+
+    def find_all_offer_category_offer_ids(self, id):
+        result = self._call('findAllOfferCategoryOfferIds', id=id, single_result=False, raw=True)
+        return [int(i) for i in result]  # HO returns ID's as strings, cast to int for convenience
```
Comment (Stranger6667) [inline pyoffers/models/application.py]: Please, remove casting. It makes the client behavior inconsistent. In some places it is casted, in some - not. Let's not introduce extra behavior on the top of HO.

**Review by Stranger6667 — COMMENTED** [review]
Please, add a changelog entry

File: pyoffers/models/core.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def update(self, **kwargs):
 
     @generic_method
     def delete(self, **kwargs):
-        return self._manager.delete(self.id, status='deleted')
+        return self._manager.update(self.id, status='deleted')
```
Comment (Stranger6667) [inline pyoffers/models/core.py]: It should be done in a separate PR, since it is a bug fix . Change log entry should be in Fixed section

File: pyoffers/models/application.py (hunk 1 above)
Comment (iamanikeev) [reply pyoffers/models/application.py]: done

File: pyoffers/models/core.py (hunk 2 above)
Comment (iamanikeev) [reply pyoffers/models/core.py]: This bug is not present in master, seems like I've put it here myself in this PR

Comment (iamanikeev) [reply pyoffers/models/core.py]: yet, I don't see it in a diff of commit where the line has changed

Comment (Stranger6667) [reply pyoffers/models/core.py]: Then it should be reverted

File: docs/changelog.rst [hunk 3]
```diff
@@ -6,6 +6,9 @@ Changelog
 `Unreleased`_
 -------------
 
+- Introduce Application model
```
Comment (Stranger6667) [inline docs/changelog.rst]: In should be under `Added` section

File: pyoffers/models/application.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from pyoffers.utils import Filter
+
+from .core import Model, ModelManager
+
+
+class Application(Model):
```
Comment (Stranger6667) [inline pyoffers/models/application.py]: Is it required?

File: tests/models/test_application.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# coding: utf-8
+CASSETTE_NAME = 'application'
+
+
+def test_find_all_offer_categories(api):
+    assert isinstance(api.application.find_all_offer_categories(), list)
```
Comment (Stranger6667) [inline tests/models/test_application.py]: could you add an assert for category class?

File: tests/models/test_offer.py [hunk 6]
```diff
@@ -53,6 +53,10 @@ def test_add_category_success(offer):
     assert offer.add_category(1) is True
 
 
+def test_get_categories(offer):
+    assert len(offer.get_categories()) == 1
```
Comment (Stranger6667) [inline tests/models/test_offer.py]: And here, please. An assert for category class

File: pyoffers/models/core.py [hunk 7]
```diff
@@ -77,7 +86,14 @@ def init_instance(self, data):
         return self.model(manager=self, **data)
 
     def _call(self, method, **kwargs):
-        return self.api._call(self.model.__name__, method, **kwargs)
+        return self.api._call(self.target, method, **kwargs)
+
+    @property
+    def target(self):
+        if self.is_application_manager:
```
Comment (Stranger6667) [inline pyoffers/models/core.py]: I propose to remove ``is_application_manager`` attribute and use inheritance instead. ``ApplicationManager`` could be placed in ``core.py`` and used as a base class for OfferCategoryManager.

Also, it would be easier to redefine target property later if it will just return ``self.model.__name__``. ``ApplicationManager``, for example, will have ``target = 'Application'``

File: pyoffers/models/core.py [hunk 8]
```diff
@@ -65,10 +66,18 @@ class ModelManager(metaclass=SelectiveInheritanceMeta):
     """
     Proxy for API methods with predefined model.
     Used to structure API. Every manager works only with single model methods.
+
+    Class properties:
+        - model: model class
+        - name: a name by which a manager will be referenced in API instance
+        - generic_methods: what methods to inherit from bas ``ModelManager``
```
Comment (Stranger6667) [inline pyoffers/models/core.py]: ``bas`` -> ``base`` or ``basic``

File: pyoffers/models/constants.py [hunk 9]
```diff
@@ -0,0 +1 @@
+APPLICATION_CONTROLLER_NAME = 'Application'
```
Comment (Stranger6667) [inline pyoffers/models/constants.py]: Seems to be redundant, because of the comment below.

File: pyoffers/models/application.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+from pyoffers.utils import Filter
```
Comment (Stranger6667) [inline pyoffers/models/application.py]: ``from ..utils import Filter``

**Review by Stranger6667 — CHANGES_REQUESTED** [review]

File: pyoffers/models/application.py (hunk 4 above)
Comment (iamanikeev) [reply pyoffers/models/application.py]: Yes. Otherwise how would I add a manager?

File: pyoffers/models/core.py (hunk 7 above)
Comment (iamanikeev) [reply pyoffers/models/core.py]: done

File: pyoffers/models/core.py (hunk 8 above)
Comment (iamanikeev) [reply pyoffers/models/core.py]: done

File: pyoffers/models/constants.py (hunk 9 above)
Comment (iamanikeev) [reply pyoffers/models/constants.py]: removed

File: pyoffers/models/application.py (hunk 10 above)
Comment (iamanikeev) [reply pyoffers/models/application.py]: this is no longer actual since this model has moved to core

**Review by Stranger6667 — APPROVED** [review]
